### PR TITLE
cmd/sync: add 'max-failure' to exit when the transfer encounters more than a certain number of errors

### LINKED ISSUE #5140: Sync tool support fail on error mode
**What would you like to be added**:

Add an option to allow sync tool to fail fast.

**Why is this needed**:
Currently producer has no feedback from workers. When workers encounter a fatal error, the producer will not stop until it has read all data, which is time consuming. It will be helpful to fail fast if any error happens.

### PR DESCRIPTION
**Pull request by zhijian-pro**
close #5140

### CODE REVIEW

**Comment by zhijian-pro** [comment]
> Maybe --max-errors or --max-failure? we have --max-failure for warmup

ok

File: pkg/sync/sync.go [hunk 1]
```diff
@@ -1282,8 +1282,55 @@ func Sync(src, dst object.ObjectStorage, config *Config) error {
...
+		}
+		return nil
+	}
+
 	if !config.Dry {
 		failed = progress.AddCountSpinner("Failed objects")
+		if config.MaxFailure > 0 {
+			go func() {
+				for {
+					if failed.Current() >= config.MaxFailure {
```
Comment (davies) [inline pkg/sync/sync.go:1324]: >

File: pkg/sync/sync.go [hunk 2]
```diff
@@ -1282,8 +1282,55 @@ func Sync(src, dst object.ObjectStorage, config *Config) error {
...
+						logger.Infof("the maximum error limit of %d was reached, triggering the quick exit process", config.MaxFailure)
+						if syncExitFunc() != nil {
```
Comment (davies) [inline pkg/sync/sync.go:1326]: will it return nil?

File: pkg/sync/sync.go [hunk 3]
```diff
@@ -1282,8 +1282,55 @@ func Sync(src, dst object.ObjectStorage, config *Config) error {
...
+				msg += fmt.Sprintf(", lost: %d", total-handled.Current())
+			}
+			logger.Info(msg)
+		} else {
+			sendStats(config.Manager)
+			logger.Debugf("This worker process has already completed its task")
+		}
+		if failed != nil {
+			if n := failed.Current(); n > 0 || total > handled.Current() {
+				return fmt.Errorf("failed to handle %d objects", n+total-handled.Current())
```
Comment (davies) [inline pkg/sync/sync.go:1313]: When failed < maxFailure, should we exit with 0?

Comment (zhijian-pro) [reply pkg/sync/sync.go:1313]: I think users expect quick results, but that doesn't mean results aren't important. So any failure should return a non-zero error code.

**Review by jiefenghuang — COMMENTED** [review]
for supporting fast failed, add context to worker may be more concise and simple
```
func worker(ctx context.Context, tasks <-chan object.Object, src, dst object.ObjectStorage, config *Config) {
    for {
        select {
            case <-ctx.Done():
                // cancel
                return
             case obj := <-tasks:
                 // origin logic
        }
        if failed >= xxx {
            ctx.cancel()
            ...
        }
    }
}
```

**Comment by zhijian-pro** [comment]
> for supporting fast failed, add context to worker may be more concise and simple
> 
> ```
> func worker(ctx context.Context, tasks <-chan object.Object, src, dst object.ObjectStorage, config *Config) {
>     for {
>         select {
>             case <-ctx.Done():
>                 // cancel
>                 return
>              case obj := <-tasks:
>                  // origin logic
>         }
>         if failed >= xxx {
>             ctx.cancel()
>             ...
>         }
>     }
> }
> ```

ok

**Review by jiefenghuang — APPROVED** [review]
LGTM

**Comment by zhijian-pro** [comment]
> for supporting fast failed, add context to worker may be more concise and simple
> 
> ```
> func worker(ctx context.Context, tasks <-chan object.Object, src, dst object.ObjectStorage, config *Config) {
>     for {
>         select {
>             case <-ctx.Done():
>                 // cancel
>                 return
>              case obj := <-tasks:
>                  // origin logic
>         }
>         if failed >= xxx {
>             ctx.cancel()
>             ...
>         }
>     }
> }
> ```

Let's do it the easy way, it's a little bit more complicated

**Pull request merged**